activation/nipost.go

```diff
@@ -377,34 +358,106 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 func (nb *NIPostBuilder) submitPoetChallenges(
 	ctx context.Context,
 	signer *signing.EdSigner,
-	deadline time.Time,
+	poetProofDeadline time.Time,
+	curPoetRoundStartDeadline time.Time,
 	challenge []byte,
-) error {
+) ([]nipost.PoETRegistration, error) {
+	// check if some registrations missing or were removed
+	nodeID := signer.NodeID()
+	registrations, err := nipost.PoetRegistrations(nb.localDB, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get poet registrations from db: %w", err)
+	}
+
+	registrationsMap := make(map[string]nipost.PoETRegistration)
+	for _, reg := range registrations {
+		registrationsMap[reg.Address] = reg
+	}
+
+	existingRegistrationsMap := make(map[string]nipost.PoETRegistration)
+	missingRegistrationsPoETAddresses := make(map[string]struct{})
+
+	for addr := range nb.poetProvers {
+		if val, ok := registrationsMap[addr]; ok {
+			existingRegistrationsMap[addr] = val
+		} else {
+			missingRegistrationsPoETAddresses[addr] = struct{}{}
+		}
+	}
+
+	existingRegistrations := maps.Values(existingRegistrationsMap)
+	if len(missingRegistrationsPoETAddresses) == 0 {
+		return existingRegistrations, nil
+	}
+
+	now := time.Now()
+
+	if curPoetRoundStartDeadline.Before(now) {
+		if len(existingRegistrations) == 0 {
+			if len(registrations) == 0 {
+				// no existing registration at all, drop current registration challenge
+				err = ErrATXChallengeExpired
+			} else {
+				// no existing registration for given poets set
+				err = ErrNoRegistrationForGivenPoetFound
+				nb.logger.Warn("revert poet configuration to previous is needed immediately",
+					zap.Error(err), log.ZShortStringer("smesherID", nodeID))
+			}
+			return nil, fmt.Errorf(
+				"%w: poet round has already started at %s (now: %s)",
+				err,
+				curPoetRoundStartDeadline,
+				now,
+			)
+		}
+		return existingRegistrations, nil
+	}
+
+	// send registrations to missing addresses
 	signature := signer.Sign(signing.POET, challenge)
 	prefix := bytes.Join([][]byte{signer.Prefix(), {byte(signing.POET)}}, nil)
-	nodeID := signer.NodeID()
-	g, ctx := errgroup.WithContext(ctx)
-	errChan := make(chan error, len(nb.poetProvers))
+
+	submitCtx, cancel := context.WithDeadline(ctx, curPoetRoundStartDeadline)
+	defer cancel()
+
+	g, ctx := errgroup.WithContext(submitCtx)
+	errChan := make(chan error, len(missingRegistrationsPoETAddresses))
+	submittedRegistrations := make([]nipost.PoETRegistration, 0)
+
 	for _, client := range nb.poetProvers {
-		g.Go(func() error {
-			errChan <- nb.submitPoetChallenge(ctx, nodeID, deadline, client, prefix, challenge, signature)
-			return nil
-		})
+		if _, ok := missingRegistrationsPoETAddresses[client.Address()]; ok {
+			g.Go(func() error {
+				registration, err := nb.submitPoetChallenge(ctx, nodeID, poetProofDeadline, client, prefix, challenge, signature)
+				if err == nil {
+					submittedRegistrations = append(submittedRegistrations, registration)
```

Comment:
The `submittedRegistrations` needs protection to be modified from many goroutines.